functorch.grad support for autograd.Function

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90418
* #90415
* #90077
* #90240
* #90037
* __->__ #89860
* #89859
* #89858

Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

zou3519 committed:
functorch.grad support for autograd.Function

Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

[ghstack-poisoned]

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/functorch/ADInterpreters.cpp:
```diff
@@ -44,12 +44,17 @@ Tensor materializeGradWrappers(const Tensor& tensor, int64_t current_level) {
   return makeTensorWrapper(tensor, current_level, /*is_immutable=*/true);
 }
 
+static Tensor base_lift(const Tensor& tensor) {
+  auto tensor_ = unwrapIfDead(tensor);
+  return materializeGradWrappers(tensor_, level());
+}
```
This is a bug-fix. The equivalent of Interpreter.lift in C++ unwraps dead tensors and then materializes grad wrappers.

zou3519 commented on torch/_functorch/autograd_function.py:
```diff
@@ -0,0 +1,196 @@
+        return autograd_function.apply(*args[1:], **kwargs)
+
+
+# "custom_function_call"
```
Naming question: I refer to autograd.Function as "AutogradFunction" sometimes, and sometimes as "CustomFunction". What do you guys want me to call it?

soulitzer replied:
Should the instance of CustomFunctionPyOperator be private? I see it being called as part of "apply', but is there a case when someone would someone call it directly?

zou3519 replied:
Everything inside `torch/_functorch/autograd_function.py` is private, including custom_function_call. Users should not be calling it directly.

Developers should interact with custom_function_call directly, instead of CustomFunctionPyOperator if that is the question. custom_function_call is an instance of a PyOperator, CustomFunctionPyOperator is just a class.

zou3519 commented on torch/autograd/function.py:
```diff
@@ -409,6 +363,68 @@ def jvp(ctx: Any, *grad_inputs: Any) -> Any:
+        if not torch._C._is_autograd_function_extension_enabled():
+            return super().apply(*args, **kwargs)
+
+        # TODO: fix circular import
+        from torch._functorch.autograd_function import custom_function_call
```
I would prefer to handle this in a follow-up:
- due to the feature flag this is not exercised in the fast-path
- It might involve a refactor of torch/_ops.py

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

[ghstack-poisoned]

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/autograd_function_db.py:
```diff
@@ -0,0 +1,250 @@
+from torch.testing._internal.common_dtype import all_types_and
+import numpy as np
+
+# Note: [autograd.Function db]
+#
+# This is a collection of autograd.Function test cases written as OpInfos
+# so they can easily be consumed by OpInfo-based tests to check if a subsystem
+# supports autograd.Function.
```
cc @mruberry @ngimel -- In this PR, we've added some autograd.Function examples as OpInfos. I have not made this available by default in the `op_db` that all tests use, but send them through functorch tests because that's the easiest way to test the autograd.Function <-> functorch interaction. Should we make them available by default?

I haven't made them available by default because autograd.Function aren't really "operators", though I can see this being useful beyond just functorch testing.

zou3519 commented:
There appear to be some issues with dynamo, but other than that, this should be ready for review

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/builder.py:
```diff
@@ -443,11 +443,12 @@ def index_source(key):
-            isinstance(value, types.BuiltinFunctionType)
-            and type(getattr(value, "__self__", None))
+            type(getattr(value, "__self__", None))
             is torch.autograd.function.FunctionMeta
             and getattr(value, "__name__", "") == "apply"
             and value == getattr(value.__self__, "apply", None)
+            and hasattr(value, "__func__")
+            and value.__func__ is torch.autograd.Function.apply.__func__
```
Note on dynamo changes:
- this PR modifies autograd.Function's apply method. It is no longer a "BuiltinFunctionType" (i.e. it is no longer a C++ function).
- AFAICT dynamo here checks if we have an autograd.Function that does not have a user-overiden apply method.
- So I modified this to check that the apply function is indeed not overiden

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

soulitzer commented on aten/src/ATen/functorch/ADInterpreters.cpp:
```diff
@@ -44,12 +44,17 @@ Tensor materializeGradWrappers(const Tensor& tensor, int64_t current_level) {
 
 Tensor JvpInterpreterPtr::lift(const Tensor& tensor) const {
-  return materializeGradWrappers(tensor, level());
+  return base_lift(tensor, level());
```
When does this happen? Why didn't we need it before?

zou3519 replied:
This is necessary to support functorch.vjp. functorch.vjp constructs a graph using GradWrapperTensors which then all become dead tensor wrappers.

In the C++ the functorch grad transform always unwraps dead tensor wrappers (https://github.com/pytorch/pytorch/blob/a6caa9c54bb5fce93a77e7407fd7e8993a767527/aten/src/ATen/functorch/ADInterpreters.cpp#L60-L68, it happens inside checkForInvalidMutationOnCaptures).

In Python, we want to be consistent with that behavior (https://github.com/pytorch/pytorch/blob/a6caa9c54bb5fce93a77e7407fd7e8993a767527/torch/_functorch/pyfunctorch.py#L107-L110). pyfunctorch today just materializes the wrappers but does not actually unwrap dead wrappers, so this PR fixes it.

Previously we didn't need it because I didn't actually test functorch.vjp; this PR exercises the functorch vjp tests and so we hit this case.

soulitzer reviewed (commented):
Pretty cool!

zou3519 commented on torch/_functorch/autograd_function.py:
```diff
@@ -0,0 +1,193 @@
+            with torch.enable_grad(), maybe_interpreter.lower():
+                output = custom_function_call(autograd_function, *unwrapped_operands)
+
+            # autograd.Function users (and ctx.mark_dirty) expect a returned input
```
Thanks for the clarification, I didn't realize it was just for ctx.mark_dirty

zou3519 commented on torch/_functorch/autograd_function.py:
```diff
@@ -0,0 +1,193 @@
+# but is now dead since we are outside the vjp context.
+#
+# PyTorch dispatcher operations, upon seeing a dead GradTensorWrapper,
+# will automatically unwrap the GradTensorWrapper when applied.
```
Yes, this refers to the dead_tensor_wrapper_fallback.

Let's assume an input to the autograd.Function is a dead tensor wrapper.
It needs to be handled at this level because otherwise ctx.save_for_backward, on saving an input for backward, will save a dead tensor wrapper. This will mess up the autograd graph, because it'll look like the graph is disconnected from the underlying Tensor.

Contrast this to what happens in an aten pytorch operator. In an aten pytorch operator, saving Tensors for backward happens AFTER a tensor has dispatched on the autograd key. dead_tensor_wrapper_fallback runs before the Autograd key, so the VariableType kernel ends up saving the underlying Tensor instead of a dead tensor wrapper.

There are some other problems that we can run into if we don't unwrap dead tensor wrapper here. For example, it is incorrect to directly pass a dead tensor wrapper to the autograd.Function's forward pass because it may grab the .data_ptr of a dead tensor wrapper and error as a result.

> By "consistent" do you mean you want to autograd function to be consistent with normal torch operators?

Yes.

soulitzer replied:
Makes sense, thanks!

zou3519 commented:
> Should the instance of CustomFunctionPyOperator be private? I see it being called as part of "apply', but is there a case when someone would someone call it directly?

I can't comment on this directly for some reason, but here goes:

Everything in the `torch/_functorch/autograd_function.py` file is private. Users should not be calling custom_function_call or CustomFunctionPyOperator directly. autograd.Function invokes custom_function_call as an implementation detail.

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

soulitzer commented on torch/testing/_internal/autograd_function_db.py:
```diff
@@ -0,0 +1,250 @@
+
+def inplace_on_select_base(x):
+    y = x.clone()
+    y0 = Select.apply(x, 0)
```
You mean to do select on `y` not `x`? Currently you are just in-placing on an unrelated clone.

zou3519 replied:
Good catch.

In turns out this behavior is banned (views created inside autograd.Function cannot be modified outside of them), so I'll just delete this test.

zou3519 replied:
Test deleted

soulitzer commented on torch/testing/_internal/autograd_function_db.py:
```diff
@@ -0,0 +1,250 @@
+        supports_out=False,
+    ),
+    OpInfo(
+        'NumpyCubeNotComposableAutogradFunction',
```
Shouldn't this one error when we run it with higher order gradients? Since we are testing with test_vjpvjp above, should that have a failure for this one?

zou3519 replied:
Computing with pytorch autograd:
- the first order gradient doesn't require grad
- (https://gist.github.com/zou3519/40ba6fcf1dc9e4140c20841d5f433d85)

The behavior of functorch transforms on trying to compute the gradient of a quantity that doesn't require grad is that it returns a Tensor full of zeros:
```
import torch
from functorch import grad
def f(x):
  return torch.zeros_like(x)

x = torch.tensor(1.)
grad(f)(x)  # torch.tensor(0)
```

the `test_vjpvjp` test is comparing vjp+vjp to vjp+{slightly modified PyTorch autograd}, so, both results end up giving us zero.

If we want to do something better here, we either need to change how functorch's grad transform works in general (or add some options to it), or improve the ctx.once_differentiable decorator. From my testing it looks like in regular PyTorch autograd, ctx.once_differentiable causes the output to not require_grad, but it would be better if it somehow attached a CantComputeGradients grad_fn to the output.

soulitzer replied:
From reading https://github.com/pytorch/pytorch/blob/master/torch/autograd/function.py#L422-L430

It looks like it already does attach the error grad_fn to backward, but it needs at least one of the inputs to backward to require grad. Taking the code example, if we pass in a grad_out that requires_grad that seems to produce the desired error.

```python
import torch

def to_numpy(x):
    return x.numpy()

class NumpyCubeNotComposable(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input):
        input_np = to_numpy(input)

        ctx.input_np = input_np
        ctx.device = input.device

        return torch.tensor(input_np ** 3, device=input.device), input_np

    @staticmethod
    @torch.autograd.function.once_differentiable
    def backward(ctx, grad_output, grad_saved):
        result_np = 3 * (ctx.input_np ** 2)
        return torch.tensor(result_np, device=ctx.device)

x = torch.randn([], requires_grad=True)
y, _ = NumpyCubeNotComposable.apply(x)
grad_y = torch.rand_like(y, requires_grad=True)
gx, = torch.autograd.grad(y, x, create_graph=True, grad_outputs=grad_y)

gx.backward() # RuntimeError: trying to differentiate twice a function that was marked with @once_differentiable
```

zou3519 replied:
Yeah, I can add a one-off test for this

zou3519 replied:
Fun fact: once_differentiable doesn't quite work with functorch grad. The problem is that it uses torch._C._functions.DelayedError, and DelayedError is a torch::autograd::Node that doesn't handle functorch.

I've brainstormed two potential solutions:
1. make a DelayedError autograd.Function, use that instead of  torch._C._functions.DelayedError when functorch is active
2. turn DelayedError into a ATen operation. The forward pass is an identity, while the backward pass raises an error.

Thoughts? (2) seems better for performance (it involves less overhead)

soulitzer replied:
Either sounds good to me! Maybe just the easier option for now, and leave (2) as a follow up if perf becomes an issue?

zou3519 replied:
Punting resolution of this to follow-up since there is more discussion to be had. For now I added a one-off test with an expectedFailure on it

soulitzer commented on torch/testing/_internal/autograd_function_db.py:
```diff
@@ -0,0 +1,250 @@
+# Axes:
+# - saves {output, input, intermediate, non-tensor}
+# - {inputs, outputs} x {single tensor, tensors, arbitrary objects}
+# - Uses {mark_dirty, mark_non_differentiable, once_differentiable}
```
Very comprehensive testing!

For even greater completeness, we could add uses of `ctx.needs_input_grad` and `set_materialize_grads`.

zou3519 replied:
Good catch, will add

zou3519 replied:
I added some manual tests for these. They don't really affect correctness, so adding to autograd_function_db wasn't sufficient.

soulitzer commented on torch/_functorch/autograd_function.py:
```diff
@@ -0,0 +1,196 @@
+
+    # TODO: The name of the grad_fn is GeneratedBackward. This isn't a great UX,
+    # but in theory functorch users shouldn't be peeking at the grad_fn.
+    # We should try to generate a better name for this.
```
+1

zou3519 replied:
Will be handled in a follow-up. I have some ideas for how to do this, they all look pretty messy though

soulitzer reviewed (commented):
Just some small comments on testing, otherwise looks good to me

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 commented on torch/_functorch/autograd_function.py:
```diff
@@ -0,0 +1,196 @@
+    # We should try to generate a better name for this.
+    class Generated(_SingleLevelFunction):
+        @staticmethod
+        def forward(*operands):
+            unwrapped_operands = pytree.tree_map_only(
+                torch.Tensor,
+                lambda x: _unwrap_for_grad(x, level),
+                operands)
+            with torch.enable_grad(), maybe_interpreter.lower():
+                output = custom_function_call(autograd_function, *unwrapped_operands)
+
+            # autograd.Function's ctx.mark_dirty expect a returned input
+            # to have the same object identity as the input.
+            # Mode-only functorch will greatly simplify this logic.
+            return wrap_outputs_maintaining_identity(
+                output,
+                unwrapped_operands,
+                operands,
+                level)
```
@soulitzer -- It looks like autograd.Function only preserves object identity if the input has requires_grad=True. Is this intentional?

The effect of this is that when someone runs e.g. NumpyExp_ on a Tensor that doesn't require grad inside of functorch.grad, it complains that the output being returned is not the same as the input. The output is not the same as the input because this logic highlighted fails since the the output of autograd.Function does not have the same object identity as the input unless the input has requires_grad=True:

```py
import torch
import numpy as np

class NumpyExp_(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        x_np = x.numpy()
        np.exp(x_np, x_np)
        ctx.mark_dirty(x)
        ctx.save_for_backward(x)
        return x

    @staticmethod
    def backward(ctx, grad_output):
        output, = ctx.saved_tensors
        return NumpyMul.apply(grad_output, output)

x = torch.tensor(1., requires_grad=True)
z = x.clone()
y = NumpyExp_.apply(z)
print(y is z)
```

soulitzer replied:
Yeah that looks weird, I think it would be good to make that behavior consistent between using requires_grad=True/False. Should be a small fix.

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

[ghstack-poisoned]

soulitzer commented on torch/testing/_internal/autograd_function_db.py:
```diff
@@ -0,0 +1,247 @@
+        supports_out=False,
+    ),
+    OpInfo(
+        'NumpyExpAutogradFunction',
```
nit: Rename this to NumpyExpInplaceAutogradFunction or NumpyExp_AutogradFunction? So that it is more clear that is doing in-place?

zou3519 replied:
This is actually out-of-place, see `op=lambda x: NumpyExp_.apply(x.clone())` on the next line. The correct way to make an in-place OpInfo is to have an OpInfo for the out-of-place variant and then setup an "inplace_variant" for it

soulitzer replied:
Or alternatively mention MarkDirty in the name?

soulitzer commented on test/functorch/test_ops.py:
```diff
@@ -36,6 +36,11 @@
+    autograd_function_db
+)
+
+torch._C._set_autograd_function_extension_enabled(True)
```
Would it be better if we did this in a context manager or something? Like if someone runs this file then goes to run other stuff, this would unexpectedly affect the run of those programs?

zou3519 replied:
Sure. It doesn't matter most of the time, but it can matter.

run_test.py executes the tests by shelling out for each one, so there is no problem for that (the subprocess has its own pytorch). If a user ran `python test/functorch/test_ops.py` followed by `python test/test_autograd.py`, both of these happen in different processes, so that's also OK.

Now the fun starts when someone uses pytest over a folder, like `pytest test/` or multiple files. I don't know what happens here so I'll just use the context manager version of the flag and call it a day

soulitzer replied:
Interesting, I didn't think of pytest. Another situation is if someone imports something from this file?

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

[ghstack-poisoned]

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 committed:
Update on "functorch.grad support for autograd.Function"


Happy to split this PR more if it helps.

This PR adds functorch.grad support for autograd.Function. There's a lot
going on; here is the high level picture and there are more details as
comments in the code.

Mechanism (PyOperator)
- Somehow, autograd.Function needs to dispatch with functorch. This is
necessary because every layer of functorch needs to see the
autograd.Function; grad layers need to preserve the backward pass.
- The mechanism for this is via PyOperator. If functorch transforms are
active, then we wrap the autograd.Function in a `custom_function_call`
PyOperator where we are able to define various rules for functorch
transforms.
- `custom_function_call` has a rule for the functorch grad transform.

autograd.Function changes
- I needed to make some changes to autograd.Function to make this work.
- First, this PR splits autograd.Function into a _SingleLevelFunction
(that works with a single level of functorch transform) and
autograd.Function (which works with multiple levels). This is necessary
because functorch's grad rule needs some way of specifying a backward
pass for that level only.
- This PR changes autograd.Function's apply to eitehr call
`custom_function_call` (if functorch is active) or super().apply (if
functorch isn't active).

Testing
- Most of this PR is just testing. It creates an autograd.Function
OpInfo database that then gets passed to the functorch grad-based tests
(grad, vjp, vjpvjp).
- Since functorch transform tests are autogenerated from OpInfo tests,
this is the easiest way to test various autograd.Function with
functorch.

Future
- jvp and vmap support coming next
- better error message (functorch only supports autograd.Function that
have the optional setup_context staticmethod)
- documentation to come when we remove the feature flag

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

zou3519 commented:
@pytorchbot merge -f

zou3519 commented:
@pytorchbot merge -f "all test failures look unrelated (they didn't fail on the topmost PR)

zou3519 commented:
@pytorchbot merge -f "all test failures look unrelated (they didn't fail on the topmost PR)"

soulitzer commented on test/functorch/test_eager_transforms.py:
```diff
@@ -978,6 +979,114 @@ def h(x):
+        grad_y = torch.randn_like(x)
+
+        def h(x, grad_y):
+            _, vjp_fn = vjp(f, x)
```
I think you forgot to define f for this one

zou3519 replied:
Lol, looks like the expected failure masked this. Thanks for catching.

soulitzer commented on test/functorch/test_eager_transforms.py:
```diff
@@ -978,6 +979,114 @@ def h(x):
+            y, _ = NumpyCubeNotComposable.apply(x)
+            return y
+
+        # regular autograd x vjp
```
This might not be actionable, but if you instead tested with `grad`, even if you are doing a single grad that would fail by itself since you'd always be constructing the second order graph. This might be unexpected to users though.

zou3519 replied:
are you saying that grad(f)(x) would raise an error?

soulitzer replied:
Whoops nevermind, no it doesn't, because creating the graph is fine
